working/333 - shared memory multithreading/proposal.md

```diff
@@ -0,0 +1,1582 @@
+piece of Dart code you want to call on a specific thread into a function and
+then:
+
+1. Send Dart `isolateLocal` callback to be executed on the specific thread, but
+   make it enter (and leave) the target isolate.
+2. Create an isolate specific to the target thread (e.g. special _platform
+   isolate_ for running on main platform thread) and have callbacks to be run in
+   that isolate.
+
+However the issues described above equally apply here: you either hit a problem
+with stalling the caller by waiting to acquire an exclusive access to an isolate
+or you hit a problem with ergonomics around the lack of shared state.
+
+See [go/dart-interop-native-threading][] and [go/dart-platform-thread][] for
+more details around the challenge of crossing isolate-to-thread chasm and why
+all different solutions fall short.
+
+[native-callable-isolate-local]: https://api.dart.dev/stable/3.2.4/dart-ffi/NativeCallable/NativeCallable.isolateLocal.html
+[native-callable-listener]: https://api.dart.dev/stable/3.2.4/dart-ffi/NativeCallable/NativeCallable.listener.html
+[go/dart-interop-native-threading]: http://go/dart-interop-native-threading
+[go/dart-platform-thread]: http://go/dart-platform-thread
+
+## Map of the Territory
+
+Before we discuss our proposal for Dart it is worth look at what other popular
+and niche languages do around share memory multithreading. If you feel familiar
+with the space feel free to skip to [Shareable Data](#shareable-data) section.
+
+**C/C++**, **Java**, **Scala**, **Kotlin**, **C#** all have what I would call an
+unrestricted shared memory multithreading:
+
+- objects can be accessed (read and written to) from multiple threads at once
+- static state is shared between all threads
+- you can spawn new threads using core APIs, which are part of the language
+- you can execute code on any thread, even when the thread was spawned
+  externally: thread spawned by Java can execute C code and the other way around
+  thread spawned by C can execute Java code (after a little dance of _attaching
+  Java VM to a thread_).
+
+**Python** and **Ruby**, despite being scripting languages, both provide similar
+capabilities around multithreading as well (see [`Thread`][Ruby Thread] in Ruby
+and [`threading`][] library in Python). The following Ruby program will spawn 10
+threads which all update the same global variable:
+
+```ruby
+count = 0
+threads = 5.times.map do |i|
+  puts "T#{i}: starting"
+  Thread.new do
+    count += 1
+    puts "T#{i}: done"
+  end
+end
+
+threads.each { |t| t.join }
+puts "Counted to #{count}"
+```
+
+```console
+$ ruby test.rb
+T0: starting
+T1: starting
+T2: starting
+T3: starting
+T4: starting
+T1: done
+T4: done
+T0: done
+T2: done
+T3: done
+Counted to 5
+```
+
+Concurrency in both languages is severely limited by a global lock which
+protects interpreter's integrity. This lock is known _Global Interpreter Lock_
+(GIL) in Python and _Global VM Lock_ (GVL) in Ruby. GIL/GVL ensures that an
```

Comment:
Maybe also mention that this forces wrapping of existing native libraries for callbacks. I believe the python ecosystem has infra to generate native code for callbacks that basically block to acquire the GIL and then forward the callback to python.

Reply:
The Python standard library can generate wrappings at runtime. For example,

```
from ctypes import *

CMPFUNC = CFUNCTYPE(c_int, POINTER(c_int), POINTER(c_int))
def py_cmp_func(a, b):
    return a[0] - b[0]

IntArray5 = c_int * 5
ia = IntArray5(5, 1, 7, 33, 99)
cmp_func = CMPFUNC(py_cmp_func). 
qsort(ia, len(ia), sizeof(c_int), cmp_func) 
```

The ecosystem has packages with better ergonomics.